pages/posts/supporting-polymorphism.md

```diff
@@ -0,0 +1,431 @@
+
+To support this "looking inside" other keywords behavior, we introduced `unevaluatedProperties` with draft 2019-09.  `unevaluatedProperties` does exactly the same thing as `additionalProperties` except that it can "look inside" subschemas of the schema object to which it belongs.  So this schema
+
+```json
+{
+  "allOf": [
+    {
+      "properties": {
+        "foo": true
+      }
+    }
+  ],
+  "unevaluatedProperties": false
+}
+```
+
+will properly allow objects with a `foo` property because it can "see" the `foo` property declaration inside the subschema in `allOf`.
+
+So let's add `"unevaluatedProperties": false` to our schemas and see what happens.
+
+```json
+{
+  "id": "vehicle",
+  "type": "object",
+  "properties": {
+    "passengerCapacity": { "type": "integer" },
+    "make": { "type": "string" },
+    "model": { "type": "string" }
+  },
+  "unevaluatedProperties": false
+}
+
+{
+  "id": "boat",
+  "allOf": [
+    { "$ref": "vehicle" }
+  ],
+  "properties": {
+    "motorCount": { "type": "integer" },
+    "name": { "type": "string" }
+  },
+  "unevaluatedProperties": false
+}
+
+{
+  "id": "car",
+  "allOf": [
+    { "$ref": "vehicle" }
+  ],
+  "properties": {
+    "doorCount": { "type": "integer" },
+    "rangeInMiles": { "type": "number" },
+  },
+  "unevaluatedProperties": false
+}
+```
+
+Now, when you get JSON data for a boat
+
+```json
+{
+  "passengerCapacity": 4,
+  "motorCount": 2,
+  "name": "S.S. Unevaluated"
+}
+```
+
+validation... fails.  Why?
+
+It's the `vehicle#/unevaluatedProperties`.  The `vehicle` schema doesn't know about the `motorCount` or `name` properties, so the `$ref` in `boat` fails now.
+
+So maybe we just remove `unevaluatedProperties` from `vehicle`.  We know it's a base type and there will be extra properties from derive types, and we probably won't validate using that schema directly (only validate using `boat` and `car` which reference it), so it should be fine without that, right?
+
+For now, yes.  But what happens when you want to derive from `boat` or `car`?  Do you remove `unevaluatedProperties` from those, too?  Where do you stop?
+
+_**NOTE** This problem also exists when you try to link the schemas using `$dynamicRef`/`$dynamicAnchor` instead of `$ref`.  Verifying this is left as an exercise for the reader._
+
+It doesn't seem like `unevaluatedProperties` is going to be viable for this specific use case.
+
+So what now?  `unevaluatedProperties` was the best chance to support this model.  Indeed, this model is closely related to the purpose for which it was created: permitting properties to be declared in subschemas and referenced schemas.
+
+Luckily, there was another, less advertised feature that made it into draft 2019-09, and it's actually the mechanism that the JSON Schema specification uses to explain how `unevaluatedProperties` works: annotations.
+
+## Annotations and `unevaluatedProperties`
```

Comment:
There's a _much_ simpler solution that doesn't require understanding the internal annotations, which [might go away in the future anyway](https://github.com/orgs/json-schema-org/discussions/204).  I took out the `id`s because I wasn't sure whether the base or sealed ones should get the unadorned names.  Obviously you can put back in whatever makes the most sense to you.

```JSON
{
  "$defs": {
    "baseVehicle": {
      "type": "object",
      "properties": {
        "passengerCapacity": { "type": "integer" },
        "make": { "type": "string" },
        "model": { "type": "string" }
      }     
    },    
    "sealedVehicle": {
        "$ref": "#/$defs/baseVehicle",
        "unevaluatedProperties": false 
    },    
    "baseBoat": {
      "allOf": [
        { "$ref": "#/$defs/baseVehicle" }
      ],    
      "properties": {
        "motorCount": { "type": "integer" },
        "name": { "type": "string" }
      }     
    },    
    "sealedBoat": {
      "unevaluatedProperties": false 
    },    
    "baseCar": {
      "allOf": [
        { "$ref": "#/$defs/baseVehicle" }
      ],    
      "properties": {
        "doorCount": { "type": "integer" },
        "rangeInMiles": { "type": "number" }
      }     
    },    
    "sealedCar": {
      "$ref": "#/$defs/baseCar",
      "unevaluatedProperties": false 
    }
  }
}

Reply:
This is exactly how I model this sort of thing.

Because we codegen a compositional, rather than an inheritance model, we provide implicit conversion operators between the types we generate for `sealedCar` and `baseCar` so you can freely use one as the other. We also generate property accessors for the union of the properties defined on the relevant schema, so it "behaves as if" it were an inheritance relationship from the point of view of the user.

This lets us effectively support "multiple inheritance" in dotnet - you could mixin multiple `baseXXX` in your `allOf` and the generated support casting between them all, with accessors for all their properties.

There are some simple rules for merging property types from the most derived to the least derived - but if you build incompatible types one "wins" in terms of what is exposed in the most derived type (and your validation will likely by a perma-fail anyway!) If you cast to one of the 'allOf' types, you get its properties directly, of course - rather like explicit interface implementation.

Reply:
I've just generated the types for "#/$defs/sealedCar" from that schema fragment.
It gives you a good idea of what we end up with:

```
BaseCar.Conversions.Accessors.cs
BaseCar.Conversions.Operators.cs
BaseCar.cs
BaseCar.Object.cs
BaseCar.Properties.cs
BaseCar.Validate.AllOf.cs
BaseCar.Validate.cs
BaseCar.Validate.Object.cs

BaseVehicle.cs
BaseVehicle.Object.cs
BaseVehicle.Properties.cs
BaseVehicle.Validate.cs
BaseVehicle.Validate.Object.cs
BaseVehicle.Validate.Type.cs
schema.json

SealedCar.Conversions.Accessors.cs
SealedCar.Conversions.Operators.cs
SealedCar.cs
SealedCar.Object.cs
SealedCar.Properties.cs
SealedCar.Validate.cs
SealedCar.Validate.Object.cs
SealedCar.Validate.Ref.cs
```

`BaseCar` has conversions to and from `BaseVehicle`

```csharp
public static implicit operator BaseCar(Endjin.Test.BaseVehicle value) {}
public static implicit operator Endjin.Test.BaseVehicle(BaseCar value) {}
```

And property accessors for the aggregate set of properties

```csharp
    /// <summary>
    /// Gets PassengerCapacity.
    /// </summary>
    public Corvus.Json.JsonInteger PassengerCapacity
    {
        get
        {
            //...
        }
    }

    /// <summary>
    /// Gets Make.
    /// </summary>
    public Corvus.Json.JsonString Make
    {
        get
        {
            //...
        }
    }

    /// <summary>
    /// Gets Model.
    /// </summary>
    public Corvus.Json.JsonString Model
    {
        get
        {
            //...
        }
    }

    /// <summary>
    /// Gets DoorCount.
    /// </summary>
    public Corvus.Json.JsonInteger DoorCount
    {
        get
        {
            //...
        }
    }

    /// <summary>
    /// Gets RangeInMiles.
    /// </summary>
    public Corvus.Json.JsonNumber RangeInMiles
    {
        get
        {
            //...
        }
    }
```

`SealedCar` has the same set of properties, but the full set of conversions.

```csharp
    public static implicit operator SealedCar(Endjin.Test.BaseCar value) {}
    public static implicit operator Endjin.Test.BaseCar(SealedCar value) {}
    public static implicit operator SealedCar(Endjin.Test.BaseVehicle value) {}
    public static implicit operator Endjin.Test.BaseVehicle(SealedCar value) {}
```

In practice this means you can pass a `SealedCar` to a method that takes a `BaseVehicle`, polymorphically, but without using inheritance to implement polymorphism.

Reply:
[test.zip](https://github.com/json-schema-org/blog/files/9621029/test.zip)

(I've added the generated code fin case you're interested.)

Reply:
The primary complaint I see against this, is now we have 2 schemas for every type.  I don't know that it's simpler.  That's probably a matter of taste.

Reply:
One of them involves writing code, one of them does not.  That's a pretty straightforward metric.

You also really only need two for classes that are used both ways.  Particularly if you use `$id` or `$anchor` to give URIs to the classes intended for general use.  Then you can switch that schema without changing the URI used to reference it.

So if vehicle is an abstract class, you'd only have the base version of it.  If car is not expected to be extended, you would just have the `unevaluatedProperties` in it.  If you need to extend it later, you can factor out the base version.  And maybe boat is both concrete and extensible, so you'd need both version of it.  But maybe sailboat and motorboat are not extensible, so you'd only have one of each of those.

That's an example where out of 5 classes, only one of them needs both versions.  More might need both in the future, but that's an incremental cost.

Reply:
And if you chose to only provide the `unevaluatedProperties` ('sealed') version, you could later update the schema with a "base", without causing a disturbance in the force.